src/services/refactors/inlineFunction.ts

```diff
@@ -0,0 +1,389 @@
+                const members = type.symbol.members;
+                return members && members.get(targetSymbol.escapedName);
+            });
+        }
+        return calls.filter(n => checker.getSymbolAtLocation(n.expression) === targetSymbol);
+    }
+
+    function canInline(
+        checker: TypeChecker,
+        declaration: InlineableFunction,
+        usages: ReadonlyArray<CallExpression>,
+        call?: CallExpression
+    ) {
+        let anyHaveErrors = false;
+        let selectedHasErrors = false;
+        const body = declaration.body;
+        if (!body || body.statements.length === 0) {
+            anyHaveErrors = true;
+            selectedHasErrors = true;
+        }
+        if (containsProhibitedModifiers(declaration.modifiers)) {
+            anyHaveErrors = true;
+            selectedHasErrors = true;
+        }
+        const symbols = getExternalSymbolsReferencedInScope(declaration, checker);
+        if (!every(usages, u => areSymbolsAccessible(checker, symbols, u))) anyHaveErrors = true;
+        if (!call || !areSymbolsAccessible(checker, symbols, call)) selectedHasErrors = true;
+        return { allAvailable: !anyHaveErrors, selectedAvailable: !selectedHasErrors };
+    }
+
+    function containsProhibitedModifiers(modifiers?: NodeArray<Modifier>): boolean {
+        return !!modifiers && !!modifiers.find(mod =>
+            mod.kind === SyntaxKind.ExportKeyword);
+    }
+
+    function getEditsForAction(context: RefactorContext, actionName: string): RefactorEditInfo | undefined {
+        const { file, program, startPosition } = context;
+        const info = getInfo(program, file, startPosition);
+        if (!info) return undefined;
+        const { declaration, usages, selectedUsage } = info;
+
+        switch (actionName) {
+            case inlineAllActionName:
+                return { edits: getInlineAllEdits(context, declaration, usages) };
+            case inlineHereActionName:
+                return { edits: getInlineHereEdits(context, declaration, selectedUsage!) };
+            default:
+                return Debug.fail("invalid action");
+        }
+    }
+
+    function getInlineAllEdits(
+        context: RefactorContext,
+        declaration: InlineableFunction,
+        usages: ReadonlyArray<CallExpression>
+    ): FileTextChanges[] {
+        const { file } = context;
+        return textChanges.ChangeTracker.with(context, t => {
+            forEach(usages, oldNode => {
+                inlineAt(context, t, oldNode, declaration);
+            });
+            t.delete(file, declaration);
+        });
+    }
+
+    function getInlineHereEdits(
+        context: RefactorContext,
+        declaration: InlineableFunction,
+        selectedUsage: CallExpression
+    ): FileTextChanges[] {
+        return textChanges.ChangeTracker.with(context, t => {
+            inlineAt(context, t, selectedUsage, declaration);
+        });
+    }
+
+    function inlineAt(

```

Comment:
this function is quite long
but maybe it's inevitable to be long

Reply:
The problem is, if I extract any functions, I'm going to get a dozen parameters...